### ISSUE TITLE
Key for droplet agent non-existent on GPG Key servers

### ISSUE THREAD

**Issue by Teglgaard** [digitalocean/droplet-agent#34]
Hi
I am unable to install the droplet agent since the key does not exist on the keyserver. I am on Ubuntu 20.

I am assuming it has been deleted by mistake.

"The repository 'https://repos-droplet.digitalocean.com/apt/droplet-agent main InRelease' is not signed."

The full error
Failed to update apt cache: W:Updating from such a repository can't be done securely, and is therefore disabled by default., W:See apt-secure(8) manpage for repository creation and user configuration details., W:GPG error: https://repos-droplet.digitalocean.com/apt/droplet-agent main InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY AA15CFF453BE8E58, E:The repository 'https://repos-droplet.digitalocean.com/apt/droplet-agent main InRelease' is not signed."

**Comment by Teglgaard** [comment]
Hi I have run the exactly the same script and have just tried again to double check.

I still have the same issues. By checking keyservers directly for key AA15CFF453BE8E58 e.g. doing the search on http://keyserver.ubuntu.com/ does not yield anything. So maybe you have something that had been cached? It seems like the key does not exist right now.

**Comment by house-lee** [comment]
👋 Hi @Teglgaard ,

Thanks for providing more info. We do not submit the public key to any keyserver, instead, we manage the key via a keyring deb package (droplet-agent-keyring) which will place the gpg key under `/usr/share/keyrings/droplet-agent-keyring.gpg` 

Basically it's the same way as what's being recommended here: https://wiki.debian.org/DebianRepository/UseThirdParty

Can you run `sudo apt purge droplet-agent*` to completely remove any possible existing agent installation, and run `rm -f /etc/apt/sources.list.d/droplet-agent.list` to remove the stale apt repo info, and then run the install script again to see if the problem is fixed? If not, do you mind sharing the full output generated by the `install.sh` with me?

**Comment by Teglgaard** [comment]
Hi again!

I will try to paste it in code bracket though not pretty. So it is repetitive error messages based on three attempts.

I ran `sudo apt purge droplet-agent*`

With the result 
`
sudo apt purge droplet-agent*
[sudo] password for teglgaard: 
Reading package lists... Done
Building dependency tree       
Reading state information... Done
E: Unable to locate package droplet-agent*
E: Couldn't find any package by glob 'droplet-agent*'
`

Afterwards I ran as instructed `$ sudo rm -f /etc/apt/sources.list.d/droplet-agent.list`
rm does not yield a console message


Below is the full log for running install.sh
`
wget -qO- https://repos-droplet.digitalocean.com/install.sh | sudo bash
[sudo] password for teglgaard: 
Verifying machine compatibility...
OK
Verifying compatibility with script...
OK
Checking architecture support...
OK
Installing Droplet Agent, 1 attempt
Setting up droplet-agent apt repository...
Checking dependencies for installing droplet-agent
Importing GPG public key
Installing droplet-agent
W: GPG error: https://repos-droplet.digitalocean.com/apt/droplet-agent main InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY AA15CFF453BE8E58
E: The repository 'https://repos-droplet.digitalocean.com/apt/droplet-agent main InRelease' is not signed.
Installing Droplet Agent, 2 attempt
Setting up droplet-agent apt repository...
Checking dependencies for installing droplet-agent
Importing GPG public key
Installing droplet-agent
W: GPG error: https://repos-droplet.digitalocean.com/apt/droplet-agent main InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY AA15CFF453BE8E58
E: The repository 'https://repos-droplet.digitalocean.com/apt/droplet-agent main InRelease' is not signed.
Installing Droplet Agent, 3 attempt
Setting up droplet-agent apt repository...
Checking dependencies for installing droplet-agent
Importing GPG public key
Installing droplet-agent
W: GPG error: https://repos-droplet.digitalocean.com/apt/droplet-agent main InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY AA15CFF453BE8E58
E: The repository 'https://repos-droplet.digitalocean.com/apt/droplet-agent main InRelease' is not signed.
Installing Droplet Agent, 4 attempt
Setting up droplet-agent apt repository...
Checking dependencies for installing droplet-agent
Importing GPG public key
Installing droplet-agent
W: GPG error: https://repos-droplet.digitalocean.com/apt/droplet-agent main InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY AA15CFF453BE8E58
E: The repository 'https://repos-droplet.digitalocean.com/apt/droplet-agent main InRelease' is not signed.
Installing Droplet Agent, 5 attempt
Setting up droplet-agent apt repository...
Checking dependencies for installing droplet-agent
Importing GPG public key
Installing droplet-agent
W: GPG error: https://repos-droplet.digitalocean.com/apt/droplet-agent main InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY AA15CFF453BE8E58
E: The repository 'https://repos-droplet.digitalocean.com/apt/droplet-agent main InRelease' is not signed.
Install failed, will retry again later
`
Last mention is that there is a key file here `/usr/share/keyrings/droplet-agent-keyring.gpg`

I hope this helps you understand what is happening

**Comment by tetrambs** [comment]
> 👋 Hi @Teglgaard
> 
> Sorry for the late reply. We attempted a few times to reproduce the problem but still had no luck. I already let the team know about this issue and we will be watching closely to see if this is also happening to other customers.
> 
> If you don't mind, can you please submit a support ticket so that we can respond to your case more quickly?
> 
> Also, I kinda want to give it a last try to see if this can solve the problem. Can you run the following command to manually import the gpg key first
> `wget -qO- https://repos-droplet.digitalocean.com/gpg.key | sudo apt-key add -`
> 
> And then run `apt-get install droplet-agent` to see if the problem is resolved? If not, maybe try run `wget -qO- https://repos-droplet.digitalocean.com/install.sh | sudo bash` , I'm hoping the signature error should go away since the key has been manually imported.

This fixed the same issue I was having, got the agent installed.

**Comment by vishalg** [comment]
Hi @Teglgaard - can you check the permissions on the key file `/usr/share/keyrings/droplet-agent-keyring.gpg`? I had them at `rw-r-----` changing to `rw-r--r--` fixed it for me.

**Comment by Teglgaard** [comment]
@tetrambs and @vishalg  I got in contact with the team from Digital Ocean who were unable to help but I figured out that in the file for repo the key file to be use was hard coded. I remove this and suddenly it installed as it was supposed to. Digital ocean could not explain why the key was hard coded. Nonetheless it solved the problem.

So basically I just edited: /etc/apt/sources.list.d/droplet-agent.list and removed the hard coded key in front of the https address.

**Comment by benjaminblack** [comment]
Having the same issue. Editing droplet-agent.list did not solve the issue for me.

**Comment by long-nguyen** [comment]
I also got this error on Ubuntu 14. The Digital Ocean command does not work and retry everytime, that's bad.
I resolved by this command
`apt-get install droplet-agent
`

**Comment by pbharadiya** [comment]
> wave Hi @Teglgaard
> 
> Sorry for the late reply. We attempted a few times to reproduce the problem but still had no luck. I already let the team know about this issue and we will be watching closely to see if this is also happening to other customers.
> 
> If you don't mind, can you please submit a support ticket so that we can respond to your case more quickly?
> 
> Also, I kinda want to give it a last try to see if this can solve the problem. Can you run the following command to manually import the gpg key first `wget -qO- https://repos-droplet.digitalocean.com/gpg.key | sudo apt-key add -`
> 
> And then run `apt-get install droplet-agent` to see if the problem is resolved? If not, maybe try run `wget -qO- https://repos-droplet.digitalocean.com/install.sh | sudo bash` , I'm hoping the signature error should go away since the key has been manually imported.

This worked for me.

```
wget -qO- https://repos-droplet.digitalocean.com/gpg.key | sudo apt-key add -
```

**Comment by Pointotech** [comment]
We get these deprecation warnings on brand new droplets with nothing else installed. Seems wrong.

**Comment by guruskid** [comment]
Thanks it works for me

**Comment by igor-astsakhov** [comment]
I get the rpm installed but then when I try to access it from the web for my user - (root is disabled in ssh config), it just says Registering SSH Keys and thats it. Any idea what I did wrong?

**Comment by jorenbroekema** [comment]
All of the above did not work for me. After searching long and hard, this worked for me:

```sh
wget -qO - https://repos.insights.digitalocean.com/sonar-agent.asc | sudo apt-key add -
```

**Comment by house-lee** [comment]
@igor-astsakhov did you attempt to access the droplet by directly hitting the console button at the top right corner? That button is a shortcut for accessing under root. 
<img width="177" alt="image" src="https://user-images.githubusercontent.com/3804753/212153256-57682d23-1f10-46c6-b683-f31d138291bc.png">
If you have root disabled, you'll need to go to the `Access` tab of that droplet and specify the username to connect
<img width="761" alt="image" src="https://user-images.githubusercontent.com/3804753/212153546-e8a5582a-8658-4e5d-aa02-6a78cb262e31.png">

Can you give it a try see if it resolves?

**Comment by joaomnmoreira** [comment]
Hi!

I'm using Ubuntu 18.04.6 LTS and when doing apt update, below message started to appear:

```
Err:9 https://repos-droplet.digitalocean.com/apt/droplet-agent main InRelease
  The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 35696F43FC7DB4C2
```

For some reason public key is not available.

I obtain:
gpg /usr/share/keyrings/droplet-agent-keyring.gpg

```
gpg: keybox '/root/.gnupg/pubring.kbx' created
gpg: WARNING: no command supplied.  Trying to guess what you mean ...
pub   rsa2048 2020-09-18 [SC]
      F1D6B3A39F4AFEC32EBB366EAA15CFF453BE8E58
uid           DigitalOcean Droplet Engineering <eng-droplet@digitalocean.com>
sub   rsa2048 2020-09-18 [E]
```

/etc/apt/sources.list.d/droplet-agent.list

`deb [signed-by=/usr/share/keyrings/droplet-agent-keyring.gpg] https://repos-droplet.digitalocean.com/apt/droplet-agent main main
`

I have uninstalled, but when installing again it stops in gpg:

```
Verifying machine compatibility...
OK
Verifying compatibility with script...
OK
Checking architecture support...
OK
Installing Droplet Agent, 1 attempt
Setting up droplet-agent apt repository...
Checking dependencies for installing droplet-agent
Importing GPG public key
gpg: no valid OpenPGP data found.
Installing Droplet Agent, 2 attempt
Setting up droplet-agent apt repository...
Checking dependencies for installing droplet-agent
Importing GPG public key
gpg: no valid OpenPGP data found.
```

Anyone? Help is needed.

Thanks in advance.

Best regards,
João Moreira
